microprofile/openapi/src/main/java/io/helidon/microprofile/openapi/IndexBuilder.java

```diff
@@ -0,0 +1,159 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package io.helidon.microprofile.openapi;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.Charset;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+
+import org.jboss.jandex.Index;
+import org.jboss.jandex.IndexReader;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Indexer;
+
+/**
+ * Portable extension to allow construction of a Jandex index (to pass to
+ * SmallRye OpenAPI) from CDI if no {@code META-INF/jandex.idx} file exists on
+ * the class path.
+ */
+public class IndexBuilder implements Extension {
+
+    private static final String INDEX_PATH = "/META-INF/jandex.idx";
+
+    private static final Logger LOGGER = Logger.getLogger(IndexBuilder.class.getName());
+
+    private final boolean isIndexPresentOnClasspath;
+
+    private final Set<Class<?>> annotatedTypes = new HashSet<>();
+
+    /**
+     * Creates a new instance of the index builder.
+     *
+     * @throws IOException in case of error checking for the Jandex index file
+     */
+    public IndexBuilder() throws IOException {
+        isIndexPresentOnClasspath = checkForIndexFile();
+        if (isIndexPresentOnClasspath) {
+            LOGGER.log(Level.FINE, () -> String.format("Index file %s was located and will be used", INDEX_PATH));
+        } else {
+            LOGGER.log(Level.INFO, () -> String.format(
+                    "OpenAPI support could not locate the index file %s "
+                    + "so will build an in-memory index. This slows your app start-up.%n"
+                    + "Consider using the jandex maven plug-in to build the index "
+                    + "and add it to your app at build-time",
+                    INDEX_PATH));
+        }
+    }
+
+    /**
+     * Records each type that is annotated unless a Jandex index was found on
+     * the classpath (in which case we do not need to build our own in memory).
+     *
+     * @param <X> annotated type
+     * @param event {@code ProcessAnnotatedType} event
+     */
+    <X> void processAnnotatedType(@Observes ProcessAnnotatedType<X> type) {
+        if (isIndexPresentOnClasspath) {
+            return;
+        }
+        Class<?> c = type.getAnnotatedType().getJavaClass();
+        annotatedTypes.add(c);
```

Comment:
No action item here, but be aware as we discussed that the `ProcessAnnotatedType` event is fired during type discovery, and by the time you get to bean discovery the ecosystem may no longer do anything interesting with this type, i.e. it may be effectively orphaned.

Reply:
Yes. Presumably this is no different from using a pre-constructed index which might contain types that, during CDI processing, get discarded.

Reply:
Yes; good point.